src/Http/Headers/src/HttpHeaderParser.cs

```diff
@@ -27,8 +28,9 @@ public bool SupportsMultipleValues
         // pointing to the next non-whitespace character after a delimiter. E.g. if called with a start index of 0
         // for string "value , second_value", then after the call completes, 'index' must point to 's', i.e. the first
         // non-whitespace after the separator ','.
-        public abstract bool TryParseValue(StringSegment value, ref int index, out T parsedValue);
+        public abstract bool TryParseValue(StringSegment value, ref int index, [MaybeNull] out T parsedValue);
 
+        [return: MaybeNull]
```

Comment:
Why is this marked `[return: MaybeNull]`? It looks like it throws if it's null. If you change this you can remove `!` from some of the callsites.

Reply:
`BaseHeaderParser.TryParseValue` can return null while returning true so this could return a null value in that case. It would really have helped if the Single-value parser and the multi-value parser where separate types so you could declare their behavior in code.

Reply:
Can we just throw from this method when it would otherwise return null? Do any of the callers actually handle null appropriately? It looks like most callers would blow up with a NullReferenceException.

Reply:
Some code paths actually handle this well: https://github.com/dotnet/aspnetcore/blob/master/src/Http/Headers/src/CacheControlHeaderValue.cs#L330-L334. Like I said, you really want to separate single and mullti-valued parsers for this to be expressed correctly, but that's well beyond the scope of this PR.

Reply:
Would it be too big a big change to have the callers that expect the return value could be null to call TryParseValue instead? I get not wanting to make this much bigger though. My browser is really struggling when I go to the files tab of this PR 😆

Reply:
Removing all the null-forgiviness from our src files is certainly desirable. But a) I'm too lazy b) fixing internal implementation, which mostly works well today, has marginal utility. Maybe somebody else would want to give this a go.  😸